feat: Add proposal for KRaft liveness and readiness (#66)
strimzi/proposals: 044-kraft-liveness-readiness.md

### Ports
In KRaft mode there are two different ports that the node could be listening on.
If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is hard-coded to 9090.
If it has a broker process it should be listening on the replication port specified as the `inter.broker.listener.name` which is hard-coded to 9091.
If the node is running as a combined mode, it should be listening on both ports

### BrokerState
The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions in the "happy path" during a normal broker lifecycle are:

NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)

The other possible value is 127 which represents UNKNOWN.

### Current state metric and quorum

Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:

* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.

> If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.

Author: Good point, I've moved it into rejected alternatives because I think it is useful information to be aware of when reading the proposal

* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
* The current-state metric could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.

> * The broker nodes will not move out of the STARTING state until the quorum leader election has happened.

Author: Thanks, I've added this

## Proposal

There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
for a broker only, and finally for a combined node that has both roles.

The following statements describe the intent of the proposed probes:

* A controller is:

>     * "alive" if it has a process running.

Reviewer: If this is really the check we want why not use `ps`? Or an HTTP server running in the server for which the `bind`/`listen`/`accept` doesn't depend deeply on anything about the Kafka code?

Reviewer 2: I think we should mention this is the "minimal exception", not exactly what we will do. Otherwise, it's a little confusing when seeing "we just want to check if it has a process running" here, and then seeing "we check controller is listening on the port" below.

Author: I'm not sure what you mean by "minimal exception". I put above that these statements "describe the intent of the proposed probes". So the intent is that the controller should be marked alive if there is some sort of process running, then lower down we describe how we actually work out that a process is running. So that could be by seeing that it is listening on a port, or as Tom says by a basic `ps` check or running some other process in the container.

    * "ready" when it is ready to join the controller quorum.
* A broker is:
    * "alive" if it has a process running.
    * "ready" when it is ready to start accepting producer/consumer requests.
* A combined node is:
    * "alive" if it has a process running.

>     * "ready" when it is ready to start accepting producer/consumer requests.

Reviewer: A sentence here about this `readiness` not blocking connections from other controllers would really help clarify, I think.

### Controller only mode

The proposed probes are:

* Liveness: controller is listening on 9090
* Readiness: it is possible to make a TCP connection to port 9090 on the controller
(i.e. client socket gets into the `ESTABLISHED` TCP socket state)

### Broker only mode

The proposed probes are:

* Liveness: broker is listening on 9091
* Readiness: the BrokerState metric >=3 && != 127

**Note:** This means the brokers will not become ready until all the controllers 
are up and running.
This is similar to the current behaviour of brokers when ZooKeeper is not ready.

### Combined mode

The proposed probes are:

* Liveness: node is listening on 9090
* Readiness: the BrokerState metric >=3 && != 127

> **Note:** This means the nodes will not become ready until all the other nodes

Author: I hadn't really though through combining pure brokers and combined-mode brokers, KRaft mode definitely opens us up to all sorts of architecture choices!

are up and running.

> This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 

Author: I don't think this will be true for pure controller nodes, because they will become ready as soon as they are listening on the controller port. So that means they don't need to talk to each other until they are deemed "ready".

means the nodes will be able to communicate even if they are currently marked as not ready.

### Impact on the KafkaRoller

> The existing KafkaRoller checks whether a pod is marked as ready to determine whether it needs to be rolled. 

Author: I've expanded this section to include the change about observing the state directly

It does not take into account the state of other broker pods.
The move to KRaft mode introduces some new requirements in terms of when a pod should or should not be rolled:

* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 

> (i.e. not in pending state) before rolling the pod or waiting for it to become ready.

Author: I've added a sentence for pure controllers, the only requirements I could think of was rolling the controllers that weren't the active controller first.

The expected [BrokerState](https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/metadata/BrokerState.java) transitions in the "happy path" during a normal broker lifecycle are:

NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)

The other possible value is 127 which represents UNKNOWN.

### Current state metric and quorum

Some useful things to be aware of with the current-state metric and how the controller quorum affects the broker nodes:

* The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates whether a controller is currently a leader or follower.
If this metric shows one of these two states, this means the leader election has happened and a controller quorum has been successfully formed.
* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
* The current-state metric could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.

> * The broker nodes will not move to a RUNNING state until the quorum leader election has happened.

Author: I believe it will get stuck in the STARTING state so I've changed it to say that

## Proposal

There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
for a broker only, and finally for a combined node that has both roles.

The following statements describe the intent of the proposed probes:

* A controller is:
    * "alive" if it has a process running.
    * "ready" when it is ready to join the controller quorum.
* A broker is:
    * "alive" if it has a process running.
    * "ready" when it is ready to start accepting producer/consumer requests.
* A combined node is:
    * "alive" if it has a process running.
    * "ready" when it is ready to start accepting producer/consumer requests.

### Controller only mode

The proposed probes are:

* Liveness: controller is listening on 9090
* Readiness: it is possible to make a TCP connection to port 9090 on the controller

> (i.e. client socket gets into the `ESTABLISHED` TCP socket state)

Author: Since there is a gap between the process starting to listen on that port and actually accepting connections, the intention was that the readiness check does involve actually trying to connect. For the liveness check I guess we would stick with the existing test we've used elsewhere of:
```
netstat -lnt | grep -Eq 'tcp6?[[:space:]]+[0-9]+[[:space:]]+[0-9]+[[:space:]]+[^ ]+:9090.*LISTEN[[:space:]]*'
```

Then for the readiness we could use something like netcat couldn't we? E.g. `nc -w 5 -v localhost 9090 </dev/null; echo $?`

> (i.e. client socket gets into the `ESTABLISHED` TCP socket state)

Reviewer 3: I suspect this would create terrible mess in the broker logs ... there will be errors about TLS handshake etc. So I would be very careful about doing any `nc` stuff to it.

> (i.e. client socket gets into the `ESTABLISHED` TCP socket state)

Author: I hadn't considered the impact that might have on the controller logs, that could lead users to think they have a misbehaving client, when the actually don't. Looking at https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ControllerServer.scala there seems to be fewer actions taking place between the controller starting to listen and accepting connections, so perhaps there is minimal benefit to having separate checks. @tombentley what do you think? Would it be reasonable to have the same liveness and readiness checks for the controller? Specifically have the check be that the controller is listening on the port.

* Combined mode pods are marked as both alive and ready when the pod is listening on 9090 (no change from phase 1).

### Phase 3 - KafkaRoller is updated
* KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods.

### Phase 4
* Broker only readiness probe fully implemented to match this proposal.
* Combined mode readiness probe fully implemented to match this proposal.

## Affected/not affected projects

This change only affects the Cluster Operator and Kafka brokers.

## Compatibility

As part of this change the readiness probe for a Kafka broker in a ZooKeeper based cluster will be updated to 
continue being marked as ready when the BrokerState metric > 3 && != 127.
This means it will tolerate the PENDING_CONTROLLED_SHUTDOWN and SHUTTING_DOWN states.
This will not cause any compatibility problems because currently the KafkaAgent stops running once the metric reaches 
the RUNNING state.
The change will make the code between KRaft and ZooKeeper mode simpler and protect readiness if in future the agent is 
updated to continue running once the broker is "ready".

> ## Rejected alternatives

Reviewer 3: You probably don't need the header if you don' have any content.

Author: Good point, I've removed it

You can see how the state transitions in the [BrokerLifecycleManager](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/BrokerLifecycleManager.scala#L379), specifically that 
the broker transitions from STARTING to RECOVERY once it has caught up with the cluster metadata, then it transitions to RUNNING as long as it is not fenced.

The broker nodes cannot catch up with the cluster metadata until the controller quorum has been formed, so they will not move out of the STARTING state until the quorum leader election has happened.

## Proposal

There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
for a broker only, and finally for a combined node that has both roles.

The following statements describe the intent of the proposed probes:

* A controller is:
    * "alive" if it has a process running.
    * "ready" when it is ready to join the controller quorum.
* A broker is:
    * "alive" if it has a process running.
    * "ready" when it is ready to start accepting producer/consumer requests.
* A combined node is:
    * "alive" if it has a process running.

>     * "ready" when it is ready to start accepting producer/consumer requests.

Reviewer 4: does it mean it's already joined the controller quorum right? could we make it clearer?

Author: I've added a sentence to clarify this.

### Combined mode

The proposed probes are:

* Liveness: node is listening on the port of the first address in `controller.listener.names` (9090)
* Readiness: the BrokerState metric >=3 && != 127

**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 
means the nodes will be able to communicate even if they are currently marked as not ready.

### Impact on the KafkaRoller

The existing KafkaRoller checks whether a pod is marked as ready to determine whether it needs to be rolled. 
This relies on the fact that currently the pod being ready implies that the BrokerState is >= RUNNING.
Additionally, it does not take into account the state of other broker pods.
The move to KRaft mode introduces some new requirements in terms of when a pod should or should not be rolled.
This combined with the varying readiness checks means we need to change the way the KafkaRoller works:

* The KafkaRoller should observe the BrokerState metric and any other metrics it needs directly, rather than inferring the state based on readiness.

> * If more than one controller pod has not become ready, KafkaRoller should try to role the controllers that aren't the current active controller first.

Author: Good spot, fixed

* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
(i.e. not in pending state) before rolling the pod or waiting for it to become ready.

The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
The brokers do not move to RUNNING until a quorum has been formed.
For example in combined mode during normal startup the following would happen:

* All the pods are started at the same time
* The controller processes in each pod form a quorum

> * The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to STARTED and are marked as ready

Reviewer 4: is `STARTED` actually the `RUNNING` state?

Author: Yes good spot, I've updated it

* All pods are now ready

However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:

* The first pod is scheduled
* The controller process starts up but cannot form a quorum because the other controllers are missing
* The broker process starts up but since there is no controller quorum it does not move to STARTED, it stays in STARTING. This means the pod is not marked as ready

> * Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff

Reviewer 4: AFAIU, the above four lines sound to be an explanation of why the Readiness for "Broker only mode" and "Combined mode" is `the BrokerState metric >=3 && != 127`. If I got it right, would it make sense to provide this explanation in the previous section?

Author: The above four lines explains why the KafkaRoller should verify that the controller pods are not stuck in pending before waiting for broker pods to be ready. So I think it is right that it sits here. This proposal states that the KafkaRoller should stop relying on readiness to determine whether to roll pods and use it's own checks. So I think having this further up where we are discussing the readiness checks would muddle those back up again.

## Proposed order of tasks to complete this proposal

Since changes to the KafkaRoller and support for non-combined mode is needed to fully implement this proposal, the following phases of development are proposed:

### Phase 1 - Strimzi only supports combined mode
* Combined mode pods are marked as both alive and ready when the pod is listening on 9090. 
This is an improvement on the current state where the pod is marked as alive and ready as soon as it starts up.
* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.

### Phase 2 - Strimzi adds support for broker only and controller only modes
* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
* Broker only pods are marked as both alive and ready if listening on port 9091.
* Combined mode pods are marked as both alive and ready when the pod is listening on 9090 (no change from phase 1).

### Phase 3 - KafkaRoller is updated

> * KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods.

Reviewer 2: I don't think there is any dependency between "supporting broker-only and controller-only mode" and "KafkaRoller updated", right? Maybe we should note that the phase 2 and phase 3 is not strictly in this order?

Author: I think there is a dependency. The KafkaRoller cannot be updated to understand the difference between broker and controller nodes until they are supported as part of Strimzi. These two phases could be done together in the same PR maybe but I don't think it makes sense to do phase 3 and then phase 2 as you would have to make guesses about how phase 2 would be implemented.

Reviewer 2: Make sense. Thanks for the explanation.

# Liveness and Readiness probes in a KRaft Kafka cluster

This proposal describes the liveness and readiness probes that should be put in place 
for a Kafka cluster that is using KRaft rather than ZooKeeper. This includes a KRaft 
cluster with combined nodes and one that has separate controller and broker nodes.

## Current situation

These are the current probes used in Strimzi when ZooKeeper is present:

|Component| Liveness | Readiness |
|---|---|---|
| ZooKeeper | "imok" from 127.0.0.1:12181 | "imok" from 127.0.0.1:12181 |
| ZK-based broker | if (ready) { listening on 9091 } else { have ZK session } | if (ready) { return true } |

> The "ready" check in both cases is based on the existence of the file `../kafka-ready` which is 

Reviewer 3: Maybe you can clarify that both cases here means `Kafka liveness and readiness probe` and `ZooKeeper and Kafka` => I think it can be a little bit confusing with having both Kafka and Zoo in the Table right above.

created by the KafkaAgent when the broker state metric has value 3 (RUNNING).

## Background

This section contains some background information about the ports and metrics we could use for checks.

> ### Ports

Reviewer 3: Nit ... empty line after the title will make the readability in raw mode batter.

In KRaft mode there are two different ports that the node could be listening on.
If it has a controller process it should be listening on the port defined as the `controller.listener.names` which in Strimzi is hard-coded to 9090.
If it has a broker process it should be listening on the replication port specified as the `inter.broker.listener.name` which is hard-coded to 9091.
If the node is running as a combined mode, it should be listening on both ports

> ### BrokerState

Reviewer 3: Nit ... empty line after the title will make the readability in raw mode batter.

### Controller only mode

The proposed probes are:

* Liveness: controller is listening on the port of the first address in `controller.listener.names` (9090)
* Readiness: controller is listening on the port of the first address in `controller.listener.names` (9090)

### Broker only mode

The proposed probes are:

* Liveness: broker is listening on the port of the address of `inter.broker.listener.name` (9091)
* Readiness: the BrokerState metric >=3 && != 127

**Note:** This means the brokers will not become ready until a majority of the controllers 
are up and running.
This is similar to the current behaviour of brokers when ZooKeeper is not ready.

### Combined mode

The proposed probes are:

* Liveness: node is listening on the port of the first address in `controller.listener.names` (9090)
* Readiness: the BrokerState metric >=3 && != 127

**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 

> means the nodes will be able to communicate even if they are currently marked as not ready.

Reviewer 3: The part `is listening on the port of the first address in` in this section is a bit confusing. Neither `controller.listener.names` nor `inter.broker.listener.name` contain any address the broker is listening to. They contain the listener name. It is also not configurable anyway - neither the name and the port.

So this leaves me a bit confused. Does the way it is specified here mean some sophisticated way to parse out the listener address fro the configuration? Or is it really just an overcomplicated thing to say that you will grep netstat output for listening on port 9090 / 9091? If the second, please simplify it to make it a bit more clear what does it mean.

Author: This was a change that was proposed by @tombentley, originally it just stated the port directly as you suggest. My assumption was that in reality the implementation would just check for port 9090/9091, however I thought it was useful in the proposal to be specific that those ports actually represent e.g. the port for the controller listener, or the port for the inter broker listener.

Reviewer 3: I think it sounds weird and confusing this way. First, there is no address, and second, it might be read as that you do some parsing of the configs. Maybe you can then say something like this?

```
* Liveness: node is listening on the port 9090 which is configured in `controller.listener.names`
```

or something like that?

Author: Good suggestion, I've updated it

> ### Impact on the KafkaRoller

Reviewer 3: From here on, the proposal seems to be quite confusing in terms of what is proposes, what is being approved and what should be done in the future. I think that need to be clarified and we need to make sure there is a clear deliverable (or make it clear that there is no real deliverable).

It seems to me that, this proposal:
* Suggests how the liveness and readiness should be implemented in phase 4
* Leaves the phase 3 completely unspecified, so essentially this proposal cannot be done until the phase 4 is specified in some proposal and approved?
* The content of the `Impact on the KafkaRoller` chapter makes sense in general, but it seems IMHO too simplified to take it as something to be approved. So I think it should have some clear note that this will be covered in detail in another proposal? (at least that is what I assume?)

----

To be clear, I do not have problem with having this agreed upon before the other parts are done. It should be just more clear what are we agreeing upon and make it clear what is just outline of something what will be discussed and approved in separate proposal (e.g. the `Impact on the KafkaRoller` part).

Author: I see what you mean, I've updated the wording to try to make it clearer what is part of this proposal and what parts are more for information to help inform a separate proposal for KafkaRoller in KRaft mode. I've also merged phase 3 and 4 below, since the KafkaRoller changes wouldn't directly be covered by this proposal, it is just required before the final changes can be made to the liveness and readiness checks. Hopefully that helps clarify.

* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff

## Proposed order of tasks to complete this proposal

Since changes to the KafkaRoller and support for non-combined mode is needed to fully implement this proposal, the following phases of development are proposed:

### Phase 1 - Strimzi only supports combined mode
* Combined mode pods are marked as both alive and ready when the pod is listening on 9090. 
This is an improvement on the current state where the pod is marked as alive and ready as soon as it starts up.
* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.

### Phase 2 - Strimzi adds support for broker only and controller only modes
* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
* Broker only pods are marked as both alive and ready if listening on port 9091.

> * Combined mode pods are marked as both alive and ready when the pod is listening on 9090 (no change from phase 1).

Reviewer 3: These will likely be shell scripts -> so do they need to wait for the actual support for the broker and controller modes? OR can they be added right away just not used for the time being?

Author: The scripts could be written in advance, however these phases are meant to describe the actual checks that can be used, so if we did write them in advance they would have to be disabled.

I think the advantage of waiting is we will need a mechanism to determine whether the node is running in broker only or controller only or combined mode, and it might be that once we have added support for those modes it becomes clearer what mechanism that is. For example you could check the server properties, but it might be we end up needing an env var on the pod, and if so it would be simpler to use that.

NOT_RUNNING (0) -> STARTING (1) -> RECOVERY (2) -> RUNNING (3) -> PENDING_CONTROLLED_SHUTDOWN (6) -> SHUTTING_DOWN (7)

The other possible value is 127 which represents UNKNOWN.

You can see how the state transitions in the [BrokerLifecycleManager](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/BrokerLifecycleManager.scala#L379), specifically that 
the broker transitions from STARTING to RECOVERY once it has caught up with the cluster metadata, then it transitions to RUNNING as long as it is not fenced.

The broker nodes cannot catch up with the cluster metadata until the controller quorum has been formed, so they will not move out of the STARTING state until the quorum leader election has happened.

## Proposal

There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
for a broker only, and finally for a combined node that has both roles.

The following statements describe the intent of the proposed probes:

* A controller is:
    * "alive" if it has a process running.

>     * "ready" when it is ready to join the controller quorum.

Reviewer 4: just to be sure I got this right. Isn't "ready" when it joined a controller quorum? or actually when it's ready to join but not joined yet? I would expect the former, or? Maybe just a matter of how it sounds in English vs Italian? :-)

Reviewer 4: Ok, makes sense now, thanks.

The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
The brokers do not move to RUNNING until a quorum has been formed.
For example in combined mode during normal startup the following would happen:

* All the pods are started at the same time
* The controller processes in each pod form a quorum
* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to RUNNING and are marked as ready
* All pods are now ready

However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:

* The first pod is scheduled
* The controller process starts up but cannot form a quorum because the other controllers are missing
* The broker process starts up but since there is no controller quorum it does not move to RUNNING, it stays in STARTING. This means the pod is not marked as ready
* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff

> ## Proposed order of tasks to complete this proposal

Reviewer 4: These 3 phases are now describing the "listening" mentioning the ports directly (9090 and 9091) and not what's in the `inter.broker.listener.name` and `controller.listener.names`. Would it be better having consistency?

Author: I've edited the part higher up to refer to the ports first and then mention they are connected to those specific configurations. I think here it's fine to just refer to the ports as that is what will ultimately be in the checks.

> Since changes to the KafkaRoller and support for non-combined mode is needed to fully implement this proposal, the following phases of development are proposed:

Reviewer 3: Maybe this gives a bit more clear statement:

```suggestion
The KRaft liveness and readiness probes can be implemented in 3 different phases.
Phases 1 and 2 can be implemented solely based on this proposal.
Phase 3 can be implemented only after additional improvements of the `KafkaRoller` which will be covered by a separate proposal.
```

* A broker is:
    * "alive" if it has a process running.
    * "ready" when it is ready to start accepting producer/consumer requests.
* A combined node is:
    * "alive" if it has a process running.
    * "ready" when it is ready to start accepting producer/consumer requests.
        * This implies that the controller quorum will have been formed while the nodes are not marked as ready, since it cannot start accepting producer/consumer requests until it can talk to the leader.
        * This relies on the node still accepting incoming connections from other controller nodes even if it isn't actually marked as "ready" yet.

### Controller only mode

The proposed probes are:

* Liveness: the Java process in the container is running
* Readiness: controller is listening on the port 9090 which is configured in `controller.listener.names`

### Broker only mode

The proposed probes are:

* Liveness: broker is listening on the port 9091 which is configured in `inter.broker.listener.name`
* Readiness: the BrokerState metric >=3 && != 127

**Note:** This means the brokers will not become ready until a majority of the controllers 
are up and running.

> This is similar to the current behaviour of brokers when ZooKeeper is not ready.

Reviewer 3: Quick question ... With ZooKeeper it works today like this:
* When the broker is starting and ZooKeeper is not available (not running, not having a quorum), the broker crash-loops.
* When the broker was already running and ready (connected to ZooKeeper) and suddenly ZooKeeper loops quorum, Kafka will keep trying to reconnect, but it does not crash-loop or become not ready. It is still ready and does what it can without ZooKeeper.

How does this behave in KRaft - mainly in the second point? What will the readiness probe do here when it was running and ready and then the controller quorum is lost? Will it be the same behavior? Will it move into un-ready?

Author: From reading the KIP my understanding is that if the broker cannot contact the lead controller it will transition back into the RECOVERY state, which means yes it would move to un-ready. I think that is acceptable though, since the broker would also stop accepting client requests at that point. Whereas in ZooKeeper mode the broker could still accept client requests if it could not contact ZooKeeper. I'll do some tests this afternoon to confirm my understanding and make sure the broker state does transition back as I described.

This is the part of the KIP that I think is relevant:

> The Broker State Machine
Currently, brokers register themselves with ZooKeeper right after they start up.  This registration accomplishes two things: it lets the broker know whether it has been elected as the controller, and it lets other nodes know how to contact it.
>
> In the post-ZooKeeper world, brokers will register themselves with the controller quorum, rather than with ZooKeeper.
>
> Currently, if a broker loses its ZooKeeper session, the controller removes it from the cluster metadata.  In the post-ZooKeeper world, the active controller removes a broker from the cluster metadata if it has not sent a MetadataFetch heartbeat in a long enough time.
>
> In the current world, a broker which can contact ZooKeeper but which is partitioned from the controller will continue serving user requests, but will not receive any metadata updates.  This can lead to some confusing and difficult situations.  For example, a producer using acks=1 might continue to produce to a leader that actually was not the leader any more, but which failed to receive the controller's LeaderAndIsrRequest moving the leadership.
>
> In the post-ZK world, cluster membership is integrated with metadata updates.  Brokers cannot continue to be members of the cluster if they cannot receive metadata updates.  While it is still possible for a broker to be partitioned from a particular client, the broker will be removed from the cluster if it is partitioned from the controller.

Author: I've done some tests but haven't actually managed to see the BrokerState metric transition out of 3 once it got to 3. I tried:
 - A single controller and removing the controller
 - 3 controllers and taking down a controller to force an election
 - 3 controllers and taking down two controllers
 - 3 controllers and taking down all controllers

In all cases the BrokerState metric stayed at value 3

Reviewer 3: But did you check if you can connect with clients or not? If it indeed blocks the clients, then we probably need to transition it and it should be exposed somehow through Kafka. If the clients can keep using it with some limitations, we can probably stick with being ready as we do today.

Author: I did some more tests this morning and it seems like the clients don't get blocked, but some requests do fail. When all controllers were down any existing producer and consumer clients were able to continue producing and consuming. However any requests that included a metadata request failed, so I couldn't start a new consumer or create a new topic. However I could create a new producer client.

Reviewer 3: Ok, that sounds similar to ZK node. So I think we can keep it ready as we do today. Thanks for the clarification.

### Combined mode

The proposed probes are:

* Liveness: the Java process in the container is running
* Readiness: the BrokerState metric >=3 && != 127

**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 
means the nodes will be able to communicate even if they are currently marked as not ready.

### Implementation of the checks

To check that the Java process is running we will execute a command in the container. The proposed command is:
```shell
for proc in $(ls -1 /proc/ | grep [0-9]); do echo "$(ls -lh /proc/$proc/exe 2>/dev/null || true)" | grep -q java; done
```
This is preferable over using a tool like `ps` (`ps aux | grep -v grep | grep java`) since it does not require a change to the tools added to the container image.

> In future this could be replaced with a call to an endpoint, since other proposals discuss adding endpoints for example to expose metrics.

Reviewer 5: ```suggestion
In future this could be replaced with a call to an endpoint, since other proposals discuss adding endpoints, for example, to expose metrics.
```

* The KafkaRoller should observe the BrokerState metric and any other metrics it needs directly, rather than inferring the state based on readiness.
* If more than one controller pod has not become ready, KafkaRoller should try to roll the controllers that aren't the current active controller first.
* If a broker pod has not become ready, KafkaRoller should check the controller quorum is formed before rolling the pod.
* If a combined pod has not become ready, KafkaRoller should check that all other combined pods have been scheduled 
(i.e. not in pending state) before rolling the pod or waiting for it to become ready.

The reason for the final bullet (combined pod not being ready) might not be immediately obvious, so it is explained below.
The brokers do not move to RUNNING until a quorum has been formed.
For example in combined mode during normal startup the following would happen:

* All the pods are started at the same time
* The controller processes in each pod form a quorum
* The broker processes start up and start talking to the controller quorum so the controller unfences them and they move to RUNNING and are marked as ready
* All pods are now ready

> However, if the pods are started initially with affinity constraints so they are all in pending, then the constraint is removed, the following would happen:

Reviewer 5: ```suggestion
However, if the pods are started with affinity constraints, they are all in a pending state and the following issue arises when forming a quorum:
```

Reviewer 5: I was a bit confused by the _constraint is removed_. I assume we're saying how the final bullet is handled with affinity constraints. Are we saying this is a limitation that needs resolving with KafkaRoller first?  Maybe we should make any dependencies more explicit?

Author: Not quite, so the below bullets describe what happens if you initially have constraints that mean the pods can't be scheduled, and then later remove that constraint. The result is basically that even though the cluster should come up, it doesn't due to the way KafkaRoller works. I've made some changes to hopefully make it clearer, see what you think.

Reviewer 5: Thanks @katheris. I understand now.

* The first pod is scheduled
* The controller process starts up but cannot form a quorum because the other controllers are missing
* The broker process starts up but since there is no controller quorum it does not move to RUNNING, it stays in STARTING. This means the pod is not marked as ready
* Because this pod doesn't become ready the other pods aren't scheduled, so the pod just sits in crash loop backoff

## Proposed order of tasks to complete this proposal

The KRaft liveness and readiness probes can be implemented in 3 different phases.
Phases 1 and 2 can be implemented solely based on this proposal.
Phase 3 can be implemented only after additional improvements of the `KafkaRoller` which will be covered by a separate proposal.

> ### Phase 1 - Strimzi only supports combined mode
> * Combined mode pods are marked as both alive and ready when there is a Java process running. 
> This is an improvement on the current state where the pod is marked as alive and ready as soon as the container starts up, regardless of whether the Java process started.
> * Update the KafkaAgent to check for a broker state of >= 3 && != 127.
> This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.
> 
> ### Phase 2 - Strimzi adds support for broker only and controller only modes
> * Controller only pod liveness and readiness checks are fully implemented to match this proposal.
> * Broker only pods are marked as both alive and ready if listening on port 9091.
> * Combined mode pods are marked as both alive and ready when there is a Java process running (no change from phase 1).
> 
> ### Phase 3 - KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods
> * Broker only readiness probe fully implemented to match this proposal.
> * Combined mode readiness probe fully implemented to match this proposal.

Reviewer 3: Just formatting improvements.

```suggestion
### Phase 1 - Strimzi only supports combined mode

* Combined mode pods are marked as both alive and ready when there is a Java process running. 
  This is an improvement on the current state where the pod is marked as alive and ready as soon as the container starts up, regardless of whether the Java process started.
* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
  This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.

### Phase 2 - Strimzi adds support for broker only and controller only modes

* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
* Broker only pods are marked as both alive and ready if listening on port 9091.
* Combined mode pods are marked as both alive and ready when there is a Java process running (no change from phase 1).

### Phase 3 - KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods

* Broker only readiness probe fully implemented to match this proposal.
* Combined mode readiness probe fully implemented to match this proposal.
```

## Affected/not affected projects

This change only affects the Cluster Operator and Kafka brokers.

## Compatibility

As part of this change the readiness probe for a Kafka broker in a ZooKeeper based cluster will be updated to 
continue being marked as ready when the BrokerState metric > 3 && != 127.
This means it will tolerate the PENDING_CONTROLLED_SHUTDOWN and SHUTTING_DOWN states.
This will not cause any compatibility problems because currently the KafkaAgent stops running once the metric reaches 
the RUNNING state.
The change will make the code between KRaft and ZooKeeper mode simpler and protect readiness if in future the agent is 
updated to continue running once the broker is "ready".

## Rejected alternatives

> ### Use the current-state metric for controller readiness
> The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates the current state of the quorum.
> Previous versions of this proposal discussed whether we could check if this metric had a state of either follower or leader and use that to infer that the controller quorum had been formed and the controller should be marked as ready.
> This approach had a couple of problems:
> * The individual state of the controller doesn't always imply the full quorum state. 
> For example, this particular controller might be offline, but the rest of the quorum is still healthy.
> Also, during a leadership election the metric might show a state of `candidate` or `voted`.
> * In general readiness checks in Kubernetes should be based on individual nodes, not the cluster as a whole.
> * The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
> Although it could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.
> * The broker nodes will not move out of the STARTING state until the quorum leader election has happened.

Reviewer 3: Just a formatting change

```suggestion
### Use the current-state metric for controller readiness

The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates the current state of the quorum.
Previous versions of this proposal discussed whether we could check if this metric had a state of either follower or leader and use that to infer that the controller quorum had been formed and the controller should be marked as ready.
This approach had a couple of problems:
* The individual state of the controller doesn't always imply the full quorum state. 
  For example, this particular controller might be offline, but the rest of the quorum is still healthy.
  Also, during a leadership election the metric might show a state of `candidate` or `voted`.
* In general readiness checks in Kubernetes should be based on individual nodes, not the cluster as a whole.
* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
  Although it could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.
* The broker nodes will not move out of the STARTING state until the quorum leader election has happened.
```
